// rclnodejs/generated/bond/bond__msg__Status.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const StatusRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  id: primitiveTypes.string,
  instance_id: primitiveTypes.string,
  active: primitiveTypes.bool,
  heartbeat_timeout: primitiveTypes.float32,
  heartbeat_period: primitiveTypes.float32,
});
const StatusRefArray = ArrayType(StatusRefStruct);
const StatusRefStructArray = StructType({
  data: StatusRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class StatusWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new StatusRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._idIntialized = true;
      this._wrapperFields.id =  new StringWrapper(other._wrapperFields.id);
      this._instance_idIntialized = true;
      this._wrapperFields.instance_id =  new StringWrapper(other._wrapperFields.instance_id);
      this._activeIntialized = true;
      this._heartbeat_timeoutIntialized = true;
      this._heartbeat_periodIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new StatusRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._idIntialized = false;
    this._wrapperFields.id = new StringWrapper();
    this._instance_idIntialized = false;
    this._wrapperFields.instance_id = new StringWrapper();
    this._activeIntialized = false;
    this._heartbeat_timeoutIntialized = false;
    this._heartbeat_periodIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new StatusWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new StatusArrayWrapper;
  }
  static get ArrayType() {
    return StatusArrayWrapper;
  }
  static get refObjectArrayType() {
    return StatusRefStructArray
  }
  static get refObjectType() {
    return StatusRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._idIntialized) {
        throw new TypeError('Invalid argument: id in Status');
      }
      if (!this._instance_idIntialized) {
        throw new TypeError('Invalid argument: instance_id in Status');
      }
      if (!this._activeIntialized) {
        throw new TypeError('Invalid argument: active in Status');
      }
      if (!this._heartbeat_timeoutIntialized) {
        throw new TypeError('Invalid argument: heartbeat_timeout in Status');
      }
      if (!this._heartbeat_periodIntialized) {
        throw new TypeError('Invalid argument: heartbeat_period in Status');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.id.freeze(own, checkConsistency);
    }
    this._refObject.id = this._wrapperFields.id.refObject;
    if (own) {
      this._wrapperFields.instance_id.freeze(own, checkConsistency);
    }
    this._refObject.instance_id = this._wrapperFields.instance_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._idIntialized = true;
    this._wrapperFields.id.data = refObject.id.data;
    this._instance_idIntialized = true;
    this._wrapperFields.instance_id.data = refObject.instance_id.data;
    this._activeIntialized = true;
    this._refObject.active = refObject.active;
    this._heartbeat_timeoutIntialized = true;
    this._refObject.heartbeat_timeout = refObject.heartbeat_timeout;
    this._heartbeat_periodIntialized = true;
    this._refObject.heartbeat_period = refObject.heartbeat_period;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.id);
    StringWrapper.freeStruct(refObject.instance_id);
  }
  static destoryRawROS(msg) {
    StatusWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'bond', subFolder: 'msg', interfaceName: 'Status'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get id() {
    if (!this._idIntialized) {
      return undefined;
    }
    return this._wrapperFields.id.data;
  }
  set id(value) {
    this._idIntialized = true;
    this._wrapperFields.id.data = value;
  }
  get instance_id() {
    if (!this._instance_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.instance_id.data;
  }
  set instance_id(value) {
    this._instance_idIntialized = true;
    this._wrapperFields.instance_id.data = value;
  }
  get active() {
    if (!this._activeIntialized) {
      return undefined;
    }
    return this._refObject.active;
  }
  set active(value) {
    this._activeIntialized = true;
    this._refObject.active = value;
  }
  get heartbeat_timeout() {
    if (!this._heartbeat_timeoutIntialized) {
      return undefined;
    }
    return this._refObject.heartbeat_timeout;
  }
  set heartbeat_timeout(value) {
    this._heartbeat_timeoutIntialized = true;
    this._refObject.heartbeat_timeout = value;
  }
  get heartbeat_period() {
    if (!this._heartbeat_periodIntialized) {
      return undefined;
    }
    return this._refObject.heartbeat_period;
  }
  set heartbeat_period(value) {
    this._heartbeat_periodIntialized = true;
    this._refObject.heartbeat_period = value;
  }
  copyRefObject(refObject) {
    this._refObject = new StatusRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._idIntialized = true;
    this._wrapperFields.id.copyRefObject(this._refObject.id);
    this._instance_idIntialized = true;
    this._wrapperFields.instance_id.copyRefObject(this._refObject.instance_id);
    this._activeIntialized = true;
    this._heartbeat_timeoutIntialized = true;
    this._heartbeat_periodIntialized = true;
  }
  copy(other) {
    this._refObject = new StatusRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._idIntialized = true;
    this._wrapperFields.id.copy(other._wrapperFields.id);
    this._instance_idIntialized = true;
    this._wrapperFields.instance_id.copy(other._wrapperFields.instance_id);
    this._activeIntialized = true;
    this._heartbeat_timeoutIntialized = true;
    this._heartbeat_periodIntialized = true;
  }
  static get classType() {
    return StatusWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"instance_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"active","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"heartbeat_timeout","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"heartbeat_period","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"bond","type":"Status","stringUpperBound":null,"isPrimitiveType":false},"msgName":"Status"};
  }
  hasMember(name) {
    let memberNames = ["header","id","instance_id","active","heartbeat_timeout","heartbeat_period"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class StatusArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof StatusWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new StatusWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatusArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to StatusArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new StatusRefArray(size);
    this._refObject = new StatusRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new StatusWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof StatusArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "StatusArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      StatusWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return StatusWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return StatusArrayWrapper;
  }
}
module.exports = StatusWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "instance_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "active",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "heartbeat_timeout",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "heartbeat_period",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "bond",
    "type": "Status",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "Status"
}
*/
