// rclnodejs/generated/rosbag2_interfaces/rosbag2_interfaces__msg__WriteSplitEvent.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const WriteSplitEventRefStruct = StructType({
  closed_file: primitiveTypes.string,
  opened_file: primitiveTypes.string,
});
const WriteSplitEventRefArray = ArrayType(WriteSplitEventRefStruct);
const WriteSplitEventRefStructArray = StructType({
  data: WriteSplitEventRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class WriteSplitEventWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new WriteSplitEventRefStruct(other._refObject.toObject());
      this._closed_fileIntialized = true;
      this._wrapperFields.closed_file =  new StringWrapper(other._wrapperFields.closed_file);
      this._opened_fileIntialized = true;
      this._wrapperFields.opened_file =  new StringWrapper(other._wrapperFields.opened_file);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new WriteSplitEventRefStruct();
    this._closed_fileIntialized = false;
    this._wrapperFields.closed_file = new StringWrapper();
    this._opened_fileIntialized = false;
    this._wrapperFields.opened_file = new StringWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new WriteSplitEventWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new WriteSplitEventArrayWrapper;
  }
  static get ArrayType() {
    return WriteSplitEventArrayWrapper;
  }
  static get refObjectArrayType() {
    return WriteSplitEventRefStructArray
  }
  static get refObjectType() {
    return WriteSplitEventRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._closed_fileIntialized) {
        throw new TypeError('Invalid argument: closed_file in WriteSplitEvent');
      }
      if (!this._opened_fileIntialized) {
        throw new TypeError('Invalid argument: opened_file in WriteSplitEvent');
      }
    }
    if (own) {
      this._wrapperFields.closed_file.freeze(own, checkConsistency);
    }
    this._refObject.closed_file = this._wrapperFields.closed_file.refObject;
    if (own) {
      this._wrapperFields.opened_file.freeze(own, checkConsistency);
    }
    this._refObject.opened_file = this._wrapperFields.opened_file.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._closed_fileIntialized = true;
    this._wrapperFields.closed_file.data = refObject.closed_file.data;
    this._opened_fileIntialized = true;
    this._wrapperFields.opened_file.data = refObject.opened_file.data;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    StringWrapper.freeStruct(refObject.closed_file);
    StringWrapper.freeStruct(refObject.opened_file);
  }
  static destoryRawROS(msg) {
    WriteSplitEventWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'rosbag2_interfaces', subFolder: 'msg', interfaceName: 'WriteSplitEvent'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get closed_file() {
    if (!this._closed_fileIntialized) {
      return undefined;
    }
    return this._wrapperFields.closed_file.data;
  }
  set closed_file(value) {
    this._closed_fileIntialized = true;
    this._wrapperFields.closed_file.data = value;
  }
  get opened_file() {
    if (!this._opened_fileIntialized) {
      return undefined;
    }
    return this._wrapperFields.opened_file.data;
  }
  set opened_file(value) {
    this._opened_fileIntialized = true;
    this._wrapperFields.opened_file.data = value;
  }
  copyRefObject(refObject) {
    this._refObject = new WriteSplitEventRefStruct(refObject.toObject());
    this._closed_fileIntialized = true;
    this._wrapperFields.closed_file.copyRefObject(this._refObject.closed_file);
    this._opened_fileIntialized = true;
    this._wrapperFields.opened_file.copyRefObject(this._refObject.opened_file);
  }
  copy(other) {
    this._refObject = new WriteSplitEventRefStruct(other._refObject.toObject());
    this._closed_fileIntialized = true;
    this._wrapperFields.closed_file.copy(other._wrapperFields.closed_file);
    this._opened_fileIntialized = true;
    this._wrapperFields.opened_file.copy(other._wrapperFields.opened_file);
  }
  static get classType() {
    return WriteSplitEventWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"closed_file","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"opened_file","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"rosbag2_interfaces","type":"WriteSplitEvent","stringUpperBound":null,"isPrimitiveType":false},"msgName":"WriteSplitEvent"};
  }
  hasMember(name) {
    let memberNames = ["closed_file","opened_file"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class WriteSplitEventArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof WriteSplitEventWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new WriteSplitEventWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WriteSplitEventArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to WriteSplitEventArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new WriteSplitEventRefArray(size);
    this._refObject = new WriteSplitEventRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new WriteSplitEventWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof WriteSplitEventArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "WriteSplitEventArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      WriteSplitEventWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return WriteSplitEventWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return WriteSplitEventArrayWrapper;
  }
}
module.exports = WriteSplitEventWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "closed_file",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "opened_file",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "rosbag2_interfaces",
    "type": "WriteSplitEvent",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "WriteSplitEvent"
}
*/
