// rclnodejs/generated/nav2_msgs/nav2_msgs__action__SmoothPath_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PathWrapper = require('../../generated/nav_msgs/nav_msgs__msg__Path.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const SmoothPath_GoalRefStruct = StructType({
  path: PathWrapper.refObjectType,
  smoother_id: primitiveTypes.string,
  max_smoothing_duration: DurationWrapper.refObjectType,
  check_for_collisions: primitiveTypes.bool,
});
const SmoothPath_GoalRefArray = ArrayType(SmoothPath_GoalRefStruct);
const SmoothPath_GoalRefStructArray = StructType({
  data: SmoothPath_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SmoothPath_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SmoothPath_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.path =  new PathWrapper(other._wrapperFields.path);
      this._smoother_idIntialized = true;
      this._wrapperFields.smoother_id =  new StringWrapper(other._wrapperFields.smoother_id);
      this._wrapperFields.max_smoothing_duration =  new DurationWrapper(other._wrapperFields.max_smoothing_duration);
      this._check_for_collisionsIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SmoothPath_GoalRefStruct();
    this._wrapperFields.path = new PathWrapper();
    this._smoother_idIntialized = false;
    this._wrapperFields.smoother_id = new StringWrapper();
    this._wrapperFields.max_smoothing_duration = new DurationWrapper();
    this._check_for_collisionsIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SmoothPath_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SmoothPath_GoalArrayWrapper;
  }
  static get ArrayType() {
    return SmoothPath_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return SmoothPath_GoalRefStructArray
  }
  static get refObjectType() {
    return SmoothPath_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._smoother_idIntialized) {
        throw new TypeError('Invalid argument: smoother_id in SmoothPath_Goal');
      }
      if (!this._check_for_collisionsIntialized) {
        throw new TypeError('Invalid argument: check_for_collisions in SmoothPath_Goal');
      }
    }
    this._wrapperFields.path.freeze(own, checkConsistency);
    this._refObject.path = this._wrapperFields.path.refObject;
    if (own) {
      this._wrapperFields.smoother_id.freeze(own, checkConsistency);
    }
    this._refObject.smoother_id = this._wrapperFields.smoother_id.refObject;
    this._wrapperFields.max_smoothing_duration.freeze(own, checkConsistency);
    this._refObject.max_smoothing_duration = this._wrapperFields.max_smoothing_duration.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.path.copyRefObject(refObject.path);
    this._smoother_idIntialized = true;
    this._wrapperFields.smoother_id.data = refObject.smoother_id.data;
    this._wrapperFields.max_smoothing_duration.copyRefObject(refObject.max_smoothing_duration);
    this._check_for_collisionsIntialized = true;
    this._refObject.check_for_collisions = refObject.check_for_collisions;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PathWrapper.freeStruct(refObject.path);
    StringWrapper.freeStruct(refObject.smoother_id);
    DurationWrapper.freeStruct(refObject.max_smoothing_duration);
  }
  static destoryRawROS(msg) {
    SmoothPath_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'SmoothPath_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get path() {
    return this._wrapperFields.path;
  }
  set path(value) {
    if (value instanceof PathWrapper) {
      this._wrapperFields.path.copy(value);
    } else {
      this._wrapperFields.path.copy(new PathWrapper(value));
    }
  }
  get smoother_id() {
    if (!this._smoother_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.smoother_id.data;
  }
  set smoother_id(value) {
    this._smoother_idIntialized = true;
    this._wrapperFields.smoother_id.data = value;
  }
  get max_smoothing_duration() {
    return this._wrapperFields.max_smoothing_duration;
  }
  set max_smoothing_duration(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.max_smoothing_duration.copy(value);
    } else {
      this._wrapperFields.max_smoothing_duration.copy(new DurationWrapper(value));
    }
  }
  get check_for_collisions() {
    if (!this._check_for_collisionsIntialized) {
      return undefined;
    }
    return this._refObject.check_for_collisions;
  }
  set check_for_collisions(value) {
    this._check_for_collisionsIntialized = true;
    this._refObject.check_for_collisions = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SmoothPath_GoalRefStruct(refObject.toObject());
    this._wrapperFields.path.copyRefObject(this._refObject.path);
    this._smoother_idIntialized = true;
    this._wrapperFields.smoother_id.copyRefObject(this._refObject.smoother_id);
    this._wrapperFields.max_smoothing_duration.copyRefObject(this._refObject.max_smoothing_duration);
    this._check_for_collisionsIntialized = true;
  }
  copy(other) {
    this._refObject = new SmoothPath_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.path.copy(other._wrapperFields.path);
    this._smoother_idIntialized = true;
    this._wrapperFields.smoother_id.copy(other._wrapperFields.smoother_id);
    this._wrapperFields.max_smoothing_duration.copy(other._wrapperFields.max_smoothing_duration);
    this._check_for_collisionsIntialized = true;
  }
  static get classType() {
    return SmoothPath_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"Path","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"smoother_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"max_smoothing_duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"check_for_collisions","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"SmoothPath_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SmoothPath_Goal"};
  }
  hasMember(name) {
    let memberNames = ["path","smoother_id","max_smoothing_duration","check_for_collisions"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SmoothPath_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SmoothPath_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SmoothPath_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SmoothPath_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SmoothPath_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SmoothPath_GoalRefArray(size);
    this._refObject = new SmoothPath_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SmoothPath_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SmoothPath_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SmoothPath_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SmoothPath_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SmoothPath_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SmoothPath_GoalArrayWrapper;
  }
}
module.exports = SmoothPath_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_msgs",
        "type": "Path",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "smoother_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "max_smoothing_duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "check_for_collisions",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "SmoothPath_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SmoothPath_Goal"
}
*/
