// rclnodejs/generated/nav2_msgs/nav2_msgs__action__ComputePathThroughPoses_Goal.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const ComputePathThroughPoses_GoalRefStruct = StructType({
  goals: PoseStampedWrapper.refObjectArrayType,
  start: PoseStampedWrapper.refObjectType,
  planner_id: primitiveTypes.string,
  use_start: primitiveTypes.bool,
});
const ComputePathThroughPoses_GoalRefArray = ArrayType(ComputePathThroughPoses_GoalRefStruct);
const ComputePathThroughPoses_GoalRefStructArray = StructType({
  data: ComputePathThroughPoses_GoalRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class ComputePathThroughPoses_GoalWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new ComputePathThroughPoses_GoalRefStruct(other._refObject.toObject());
      this._wrapperFields.goals = PoseStampedWrapper.createArray();
      this._wrapperFields.goals.copy(other._wrapperFields.goals);
      this._wrapperFields.start =  new PoseStampedWrapper(other._wrapperFields.start);
      this._planner_idIntialized = true;
      this._wrapperFields.planner_id =  new StringWrapper(other._wrapperFields.planner_id);
      this._use_startIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new ComputePathThroughPoses_GoalRefStruct();
    this._wrapperFields.goals = PoseStampedWrapper.createArray();
    this._wrapperFields.start = new PoseStampedWrapper();
    this._planner_idIntialized = false;
    this._wrapperFields.planner_id = new StringWrapper();
    this._use_startIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new ComputePathThroughPoses_GoalWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new ComputePathThroughPoses_GoalArrayWrapper;
  }
  static get ArrayType() {
    return ComputePathThroughPoses_GoalArrayWrapper;
  }
  static get refObjectArrayType() {
    return ComputePathThroughPoses_GoalRefStructArray
  }
  static get refObjectType() {
    return ComputePathThroughPoses_GoalRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._planner_idIntialized) {
        throw new TypeError('Invalid argument: planner_id in ComputePathThroughPoses_Goal');
      }
      if (!this._use_startIntialized) {
        throw new TypeError('Invalid argument: use_start in ComputePathThroughPoses_Goal');
      }
    }
    this._wrapperFields.goals.freeze(own, checkConsistency);
    this._refObject.goals = this._wrapperFields.goals.refObject;
    this._wrapperFields.start.freeze(own, checkConsistency);
    this._refObject.start = this._wrapperFields.start.refObject;
    if (own) {
      this._wrapperFields.planner_id.freeze(own, checkConsistency);
    }
    this._refObject.planner_id = this._wrapperFields.planner_id.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.goals.copyRefObject(refObject.goals);
    this._wrapperFields.start.copyRefObject(refObject.start);
    this._planner_idIntialized = true;
    this._wrapperFields.planner_id.data = refObject.planner_id.data;
    this._use_startIntialized = true;
    this._refObject.use_start = refObject.use_start;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    if (refObject.goals.size != 0) {
      PoseStampedWrapper.ArrayType.freeArray(refObject.goals);
      if (PoseStampedWrapper.ArrayType.useTypedArray) {
        // Do nothing, the v8 will take the ownership of the ArrayBuffer used by the typed array.
      } else {
        deallocator.freeStructMember(refObject.goals, PoseStampedWrapper.refObjectArrayType, 'data');
      }
    }
    PoseStampedWrapper.freeStruct(refObject.start);
    StringWrapper.freeStruct(refObject.planner_id);
  }
  static destoryRawROS(msg) {
    ComputePathThroughPoses_GoalWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'ComputePathThroughPoses_Goal'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get goals() {
    return this._wrapperFields.goals;
  }
  set goals(value) {
    this._wrapperFields.goals.fill(value);
  }
  get start() {
    return this._wrapperFields.start;
  }
  set start(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.start.copy(value);
    } else {
      this._wrapperFields.start.copy(new PoseStampedWrapper(value));
    }
  }
  get planner_id() {
    if (!this._planner_idIntialized) {
      return undefined;
    }
    return this._wrapperFields.planner_id.data;
  }
  set planner_id(value) {
    this._planner_idIntialized = true;
    this._wrapperFields.planner_id.data = value;
  }
  get use_start() {
    if (!this._use_startIntialized) {
      return undefined;
    }
    return this._refObject.use_start;
  }
  set use_start(value) {
    this._use_startIntialized = true;
    this._refObject.use_start = value;
  }
  copyRefObject(refObject) {
    this._refObject = new ComputePathThroughPoses_GoalRefStruct(refObject.toObject());
    this._wrapperFields.goals.copyRefObject(this._refObject.goals);
    this._wrapperFields.start.copyRefObject(this._refObject.start);
    this._planner_idIntialized = true;
    this._wrapperFields.planner_id.copyRefObject(this._refObject.planner_id);
    this._use_startIntialized = true;
  }
  copy(other) {
    this._refObject = new ComputePathThroughPoses_GoalRefStruct(other._refObject.toObject());
    this._wrapperFields.goals.copy(other._wrapperFields.goals);
    this._wrapperFields.start.copy(other._wrapperFields.start);
    this._planner_idIntialized = true;
    this._wrapperFields.planner_id.copy(other._wrapperFields.planner_id);
    this._use_startIntialized = true;
  }
  static get classType() {
    return ComputePathThroughPoses_GoalWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"goals","type":{"isArray":true,"arraySize":null,"isUpperBound":false,"isDynamicArray":true,"isFixedSizeArray":null,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"planner_id","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"use_start","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"ComputePathThroughPoses_Goal","stringUpperBound":null,"isPrimitiveType":false},"msgName":"ComputePathThroughPoses_Goal"};
  }
  hasMember(name) {
    let memberNames = ["goals","start","planner_id","use_start"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class ComputePathThroughPoses_GoalArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof ComputePathThroughPoses_GoalWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new ComputePathThroughPoses_GoalWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputePathThroughPoses_GoalArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to ComputePathThroughPoses_GoalArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new ComputePathThroughPoses_GoalRefArray(size);
    this._refObject = new ComputePathThroughPoses_GoalRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new ComputePathThroughPoses_GoalWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof ComputePathThroughPoses_GoalArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "ComputePathThroughPoses_GoalArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      ComputePathThroughPoses_GoalWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return ComputePathThroughPoses_GoalWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return ComputePathThroughPoses_GoalArrayWrapper;
  }
}
module.exports = ComputePathThroughPoses_GoalWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "goals",
      "type": {
        "isArray": true,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": true,
        "isFixedSizeArray": null,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "planner_id",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "use_start",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "ComputePathThroughPoses_Goal",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "ComputePathThroughPoses_Goal"
}
*/
