// rclnodejs/generated/nav2_msgs/nav2_msgs__action__NavigateThroughPoses_Feedback.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PoseStampedWrapper = require('../../generated/geometry_msgs/geometry_msgs__msg__PoseStamped.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const NavigateThroughPoses_FeedbackRefStruct = StructType({
  current_pose: PoseStampedWrapper.refObjectType,
  navigation_time: DurationWrapper.refObjectType,
  estimated_time_remaining: DurationWrapper.refObjectType,
  number_of_recoveries: primitiveTypes.int16,
  distance_remaining: primitiveTypes.float32,
  number_of_poses_remaining: primitiveTypes.int16,
});
const NavigateThroughPoses_FeedbackRefArray = ArrayType(NavigateThroughPoses_FeedbackRefStruct);
const NavigateThroughPoses_FeedbackRefStructArray = StructType({
  data: NavigateThroughPoses_FeedbackRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NavigateThroughPoses_FeedbackWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NavigateThroughPoses_FeedbackRefStruct(other._refObject.toObject());
      this._wrapperFields.current_pose =  new PoseStampedWrapper(other._wrapperFields.current_pose);
      this._wrapperFields.navigation_time =  new DurationWrapper(other._wrapperFields.navigation_time);
      this._wrapperFields.estimated_time_remaining =  new DurationWrapper(other._wrapperFields.estimated_time_remaining);
      this._number_of_recoveriesIntialized = true;
      this._distance_remainingIntialized = true;
      this._number_of_poses_remainingIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NavigateThroughPoses_FeedbackRefStruct();
    this._wrapperFields.current_pose = new PoseStampedWrapper();
    this._wrapperFields.navigation_time = new DurationWrapper();
    this._wrapperFields.estimated_time_remaining = new DurationWrapper();
    this._number_of_recoveriesIntialized = false;
    this._distance_remainingIntialized = false;
    this._number_of_poses_remainingIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new NavigateThroughPoses_FeedbackWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NavigateThroughPoses_FeedbackArrayWrapper;
  }
  static get ArrayType() {
    return NavigateThroughPoses_FeedbackArrayWrapper;
  }
  static get refObjectArrayType() {
    return NavigateThroughPoses_FeedbackRefStructArray
  }
  static get refObjectType() {
    return NavigateThroughPoses_FeedbackRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._number_of_recoveriesIntialized) {
        throw new TypeError('Invalid argument: number_of_recoveries in NavigateThroughPoses_Feedback');
      }
      if (!this._distance_remainingIntialized) {
        throw new TypeError('Invalid argument: distance_remaining in NavigateThroughPoses_Feedback');
      }
      if (!this._number_of_poses_remainingIntialized) {
        throw new TypeError('Invalid argument: number_of_poses_remaining in NavigateThroughPoses_Feedback');
      }
    }
    this._wrapperFields.current_pose.freeze(own, checkConsistency);
    this._refObject.current_pose = this._wrapperFields.current_pose.refObject;
    this._wrapperFields.navigation_time.freeze(own, checkConsistency);
    this._refObject.navigation_time = this._wrapperFields.navigation_time.refObject;
    this._wrapperFields.estimated_time_remaining.freeze(own, checkConsistency);
    this._refObject.estimated_time_remaining = this._wrapperFields.estimated_time_remaining.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.current_pose.copyRefObject(refObject.current_pose);
    this._wrapperFields.navigation_time.copyRefObject(refObject.navigation_time);
    this._wrapperFields.estimated_time_remaining.copyRefObject(refObject.estimated_time_remaining);
    this._number_of_recoveriesIntialized = true;
    this._refObject.number_of_recoveries = refObject.number_of_recoveries;
    this._distance_remainingIntialized = true;
    this._refObject.distance_remaining = refObject.distance_remaining;
    this._number_of_poses_remainingIntialized = true;
    this._refObject.number_of_poses_remaining = refObject.number_of_poses_remaining;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PoseStampedWrapper.freeStruct(refObject.current_pose);
    DurationWrapper.freeStruct(refObject.navigation_time);
    DurationWrapper.freeStruct(refObject.estimated_time_remaining);
  }
  static destoryRawROS(msg) {
    NavigateThroughPoses_FeedbackWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'NavigateThroughPoses_Feedback'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get current_pose() {
    return this._wrapperFields.current_pose;
  }
  set current_pose(value) {
    if (value instanceof PoseStampedWrapper) {
      this._wrapperFields.current_pose.copy(value);
    } else {
      this._wrapperFields.current_pose.copy(new PoseStampedWrapper(value));
    }
  }
  get navigation_time() {
    return this._wrapperFields.navigation_time;
  }
  set navigation_time(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.navigation_time.copy(value);
    } else {
      this._wrapperFields.navigation_time.copy(new DurationWrapper(value));
    }
  }
  get estimated_time_remaining() {
    return this._wrapperFields.estimated_time_remaining;
  }
  set estimated_time_remaining(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.estimated_time_remaining.copy(value);
    } else {
      this._wrapperFields.estimated_time_remaining.copy(new DurationWrapper(value));
    }
  }
  get number_of_recoveries() {
    if (!this._number_of_recoveriesIntialized) {
      return undefined;
    }
    return this._refObject.number_of_recoveries;
  }
  set number_of_recoveries(value) {
    this._number_of_recoveriesIntialized = true;
    this._refObject.number_of_recoveries = value;
  }
  get distance_remaining() {
    if (!this._distance_remainingIntialized) {
      return undefined;
    }
    return this._refObject.distance_remaining;
  }
  set distance_remaining(value) {
    this._distance_remainingIntialized = true;
    this._refObject.distance_remaining = value;
  }
  get number_of_poses_remaining() {
    if (!this._number_of_poses_remainingIntialized) {
      return undefined;
    }
    return this._refObject.number_of_poses_remaining;
  }
  set number_of_poses_remaining(value) {
    this._number_of_poses_remainingIntialized = true;
    this._refObject.number_of_poses_remaining = value;
  }
  copyRefObject(refObject) {
    this._refObject = new NavigateThroughPoses_FeedbackRefStruct(refObject.toObject());
    this._wrapperFields.current_pose.copyRefObject(this._refObject.current_pose);
    this._wrapperFields.navigation_time.copyRefObject(this._refObject.navigation_time);
    this._wrapperFields.estimated_time_remaining.copyRefObject(this._refObject.estimated_time_remaining);
    this._number_of_recoveriesIntialized = true;
    this._distance_remainingIntialized = true;
    this._number_of_poses_remainingIntialized = true;
  }
  copy(other) {
    this._refObject = new NavigateThroughPoses_FeedbackRefStruct(other._refObject.toObject());
    this._wrapperFields.current_pose.copy(other._wrapperFields.current_pose);
    this._wrapperFields.navigation_time.copy(other._wrapperFields.navigation_time);
    this._wrapperFields.estimated_time_remaining.copy(other._wrapperFields.estimated_time_remaining);
    this._number_of_recoveriesIntialized = true;
    this._distance_remainingIntialized = true;
    this._number_of_poses_remainingIntialized = true;
  }
  static get classType() {
    return NavigateThroughPoses_FeedbackWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"current_pose","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"geometry_msgs","type":"PoseStamped","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"navigation_time","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"estimated_time_remaining","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"number_of_recoveries","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"distance_remaining","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"number_of_poses_remaining","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"int16","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"NavigateThroughPoses_Feedback","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NavigateThroughPoses_Feedback"};
  }
  hasMember(name) {
    let memberNames = ["current_pose","navigation_time","estimated_time_remaining","number_of_recoveries","distance_remaining","number_of_poses_remaining"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NavigateThroughPoses_FeedbackArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NavigateThroughPoses_FeedbackWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NavigateThroughPoses_FeedbackWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavigateThroughPoses_FeedbackArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavigateThroughPoses_FeedbackArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NavigateThroughPoses_FeedbackRefArray(size);
    this._refObject = new NavigateThroughPoses_FeedbackRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NavigateThroughPoses_FeedbackWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NavigateThroughPoses_FeedbackArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NavigateThroughPoses_FeedbackArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NavigateThroughPoses_FeedbackWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NavigateThroughPoses_FeedbackWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NavigateThroughPoses_FeedbackArrayWrapper;
  }
}
module.exports = NavigateThroughPoses_FeedbackWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "current_pose",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "geometry_msgs",
        "type": "PoseStamped",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "navigation_time",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "estimated_time_remaining",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "number_of_recoveries",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "distance_remaining",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "number_of_poses_remaining",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "int16",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "NavigateThroughPoses_Feedback",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NavigateThroughPoses_Feedback"
}
*/
