// rclnodejs/generated/nav2_msgs/nav2_msgs__action__NavigateThroughPoses_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const EmptyWrapper = require('../../generated/std_msgs/std_msgs__msg__Empty.js');
const NavigateThroughPoses_ResultRefStruct = StructType({
  result: EmptyWrapper.refObjectType,
});
const NavigateThroughPoses_ResultRefArray = ArrayType(NavigateThroughPoses_ResultRefStruct);
const NavigateThroughPoses_ResultRefStructArray = StructType({
  data: NavigateThroughPoses_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class NavigateThroughPoses_ResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new NavigateThroughPoses_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.result =  new EmptyWrapper(other._wrapperFields.result);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new NavigateThroughPoses_ResultRefStruct();
    this._wrapperFields.result = new EmptyWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new NavigateThroughPoses_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new NavigateThroughPoses_ResultArrayWrapper;
  }
  static get ArrayType() {
    return NavigateThroughPoses_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return NavigateThroughPoses_ResultRefStructArray
  }
  static get refObjectType() {
    return NavigateThroughPoses_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.result.freeze(own, checkConsistency);
    this._refObject.result = this._wrapperFields.result.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.result.copyRefObject(refObject.result);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    EmptyWrapper.freeStruct(refObject.result);
  }
  static destoryRawROS(msg) {
    NavigateThroughPoses_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'NavigateThroughPoses_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get result() {
    return this._wrapperFields.result;
  }
  set result(value) {
    if (value instanceof EmptyWrapper) {
      this._wrapperFields.result.copy(value);
    } else {
      this._wrapperFields.result.copy(new EmptyWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new NavigateThroughPoses_ResultRefStruct(refObject.toObject());
    this._wrapperFields.result.copyRefObject(this._refObject.result);
  }
  copy(other) {
    this._refObject = new NavigateThroughPoses_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.result.copy(other._wrapperFields.result);
  }
  static get classType() {
    return NavigateThroughPoses_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"result","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Empty","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"NavigateThroughPoses_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"NavigateThroughPoses_Result"};
  }
  hasMember(name) {
    let memberNames = ["result"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class NavigateThroughPoses_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof NavigateThroughPoses_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new NavigateThroughPoses_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavigateThroughPoses_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to NavigateThroughPoses_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new NavigateThroughPoses_ResultRefArray(size);
    this._refObject = new NavigateThroughPoses_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new NavigateThroughPoses_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof NavigateThroughPoses_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "NavigateThroughPoses_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      NavigateThroughPoses_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return NavigateThroughPoses_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return NavigateThroughPoses_ResultArrayWrapper;
  }
}
module.exports = NavigateThroughPoses_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "result",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Empty",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "NavigateThroughPoses_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "NavigateThroughPoses_Result"
}
*/
