// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__SpeedLimit.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const SpeedLimitRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  percentage: primitiveTypes.bool,
  speed_limit: primitiveTypes.float64,
});
const SpeedLimitRefArray = ArrayType(SpeedLimitRefStruct);
const SpeedLimitRefStructArray = StructType({
  data: SpeedLimitRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SpeedLimitWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SpeedLimitRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._percentageIntialized = true;
      this._speed_limitIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SpeedLimitRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._percentageIntialized = false;
    this._speed_limitIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SpeedLimitWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SpeedLimitArrayWrapper;
  }
  static get ArrayType() {
    return SpeedLimitArrayWrapper;
  }
  static get refObjectArrayType() {
    return SpeedLimitRefStructArray
  }
  static get refObjectType() {
    return SpeedLimitRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._percentageIntialized) {
        throw new TypeError('Invalid argument: percentage in SpeedLimit');
      }
      if (!this._speed_limitIntialized) {
        throw new TypeError('Invalid argument: speed_limit in SpeedLimit');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._percentageIntialized = true;
    this._refObject.percentage = refObject.percentage;
    this._speed_limitIntialized = true;
    this._refObject.speed_limit = refObject.speed_limit;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
  }
  static destoryRawROS(msg) {
    SpeedLimitWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'SpeedLimit'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get percentage() {
    if (!this._percentageIntialized) {
      return undefined;
    }
    return this._refObject.percentage;
  }
  set percentage(value) {
    this._percentageIntialized = true;
    this._refObject.percentage = value;
  }
  get speed_limit() {
    if (!this._speed_limitIntialized) {
      return undefined;
    }
    return this._refObject.speed_limit;
  }
  set speed_limit(value) {
    this._speed_limitIntialized = true;
    this._refObject.speed_limit = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SpeedLimitRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._percentageIntialized = true;
    this._speed_limitIntialized = true;
  }
  copy(other) {
    this._refObject = new SpeedLimitRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._percentageIntialized = true;
    this._speed_limitIntialized = true;
  }
  static get classType() {
    return SpeedLimitWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"percentage","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"speed_limit","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float64","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"SpeedLimit","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SpeedLimit"};
  }
  hasMember(name) {
    let memberNames = ["header","percentage","speed_limit"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SpeedLimitArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SpeedLimitWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SpeedLimitWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SpeedLimitArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SpeedLimitArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SpeedLimitRefArray(size);
    this._refObject = new SpeedLimitRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SpeedLimitWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SpeedLimitArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SpeedLimitArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SpeedLimitWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SpeedLimitWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SpeedLimitArrayWrapper;
  }
}
module.exports = SpeedLimitWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "percentage",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "speed_limit",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float64",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "SpeedLimit",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SpeedLimit"
}
*/
