// rclnodejs/generated/nav2_msgs/nav2_msgs__msg__CostmapFilterInfo.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const HeaderWrapper = require('../../generated/std_msgs/std_msgs__msg__Header.js');
const StringWrapper = require('../../generated/std_msgs/std_msgs__msg__String.js');
const CostmapFilterInfoRefStruct = StructType({
  header: HeaderWrapper.refObjectType,
  type: primitiveTypes.uint8,
  filter_mask_topic: primitiveTypes.string,
  base: primitiveTypes.float32,
  multiplier: primitiveTypes.float32,
});
const CostmapFilterInfoRefArray = ArrayType(CostmapFilterInfoRefStruct);
const CostmapFilterInfoRefStructArray = StructType({
  data: CostmapFilterInfoRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class CostmapFilterInfoWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new CostmapFilterInfoRefStruct(other._refObject.toObject());
      this._wrapperFields.header =  new HeaderWrapper(other._wrapperFields.header);
      this._typeIntialized = true;
      this._filter_mask_topicIntialized = true;
      this._wrapperFields.filter_mask_topic =  new StringWrapper(other._wrapperFields.filter_mask_topic);
      this._baseIntialized = true;
      this._multiplierIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new CostmapFilterInfoRefStruct();
    this._wrapperFields.header = new HeaderWrapper();
    this._typeIntialized = false;
    this._filter_mask_topicIntialized = false;
    this._wrapperFields.filter_mask_topic = new StringWrapper();
    this._baseIntialized = false;
    this._multiplierIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new CostmapFilterInfoWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new CostmapFilterInfoArrayWrapper;
  }
  static get ArrayType() {
    return CostmapFilterInfoArrayWrapper;
  }
  static get refObjectArrayType() {
    return CostmapFilterInfoRefStructArray
  }
  static get refObjectType() {
    return CostmapFilterInfoRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._typeIntialized) {
        throw new TypeError('Invalid argument: type in CostmapFilterInfo');
      }
      if (!this._filter_mask_topicIntialized) {
        throw new TypeError('Invalid argument: filter_mask_topic in CostmapFilterInfo');
      }
      if (!this._baseIntialized) {
        throw new TypeError('Invalid argument: base in CostmapFilterInfo');
      }
      if (!this._multiplierIntialized) {
        throw new TypeError('Invalid argument: multiplier in CostmapFilterInfo');
      }
    }
    this._wrapperFields.header.freeze(own, checkConsistency);
    this._refObject.header = this._wrapperFields.header.refObject;
    if (own) {
      this._wrapperFields.filter_mask_topic.freeze(own, checkConsistency);
    }
    this._refObject.filter_mask_topic = this._wrapperFields.filter_mask_topic.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.header.copyRefObject(refObject.header);
    this._typeIntialized = true;
    this._refObject.type = refObject.type;
    this._filter_mask_topicIntialized = true;
    this._wrapperFields.filter_mask_topic.data = refObject.filter_mask_topic.data;
    this._baseIntialized = true;
    this._refObject.base = refObject.base;
    this._multiplierIntialized = true;
    this._refObject.multiplier = refObject.multiplier;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    HeaderWrapper.freeStruct(refObject.header);
    StringWrapper.freeStruct(refObject.filter_mask_topic);
  }
  static destoryRawROS(msg) {
    CostmapFilterInfoWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'msg', interfaceName: 'CostmapFilterInfo'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get header() {
    return this._wrapperFields.header;
  }
  set header(value) {
    if (value instanceof HeaderWrapper) {
      this._wrapperFields.header.copy(value);
    } else {
      this._wrapperFields.header.copy(new HeaderWrapper(value));
    }
  }
  get type() {
    if (!this._typeIntialized) {
      return undefined;
    }
    return this._refObject.type;
  }
  set type(value) {
    this._typeIntialized = true;
    this._refObject.type = value;
  }
  get filter_mask_topic() {
    if (!this._filter_mask_topicIntialized) {
      return undefined;
    }
    return this._wrapperFields.filter_mask_topic.data;
  }
  set filter_mask_topic(value) {
    this._filter_mask_topicIntialized = true;
    this._wrapperFields.filter_mask_topic.data = value;
  }
  get base() {
    if (!this._baseIntialized) {
      return undefined;
    }
    return this._refObject.base;
  }
  set base(value) {
    this._baseIntialized = true;
    this._refObject.base = value;
  }
  get multiplier() {
    if (!this._multiplierIntialized) {
      return undefined;
    }
    return this._refObject.multiplier;
  }
  set multiplier(value) {
    this._multiplierIntialized = true;
    this._refObject.multiplier = value;
  }
  copyRefObject(refObject) {
    this._refObject = new CostmapFilterInfoRefStruct(refObject.toObject());
    this._wrapperFields.header.copyRefObject(this._refObject.header);
    this._typeIntialized = true;
    this._filter_mask_topicIntialized = true;
    this._wrapperFields.filter_mask_topic.copyRefObject(this._refObject.filter_mask_topic);
    this._baseIntialized = true;
    this._multiplierIntialized = true;
  }
  copy(other) {
    this._refObject = new CostmapFilterInfoRefStruct(other._refObject.toObject());
    this._wrapperFields.header.copy(other._wrapperFields.header);
    this._typeIntialized = true;
    this._filter_mask_topicIntialized = true;
    this._wrapperFields.filter_mask_topic.copy(other._wrapperFields.filter_mask_topic);
    this._baseIntialized = true;
    this._multiplierIntialized = true;
  }
  static get classType() {
    return CostmapFilterInfoWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"header","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"std_msgs","type":"Header","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"type","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"uint8","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"filter_mask_topic","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"string","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"base","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null},{"name":"multiplier","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"float32","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"CostmapFilterInfo","stringUpperBound":null,"isPrimitiveType":false},"msgName":"CostmapFilterInfo"};
  }
  hasMember(name) {
    let memberNames = ["header","type","filter_mask_topic","base","multiplier"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class CostmapFilterInfoArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof CostmapFilterInfoWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new CostmapFilterInfoWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CostmapFilterInfoArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to CostmapFilterInfoArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new CostmapFilterInfoRefArray(size);
    this._refObject = new CostmapFilterInfoRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new CostmapFilterInfoWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof CostmapFilterInfoArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "CostmapFilterInfoArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      CostmapFilterInfoWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return CostmapFilterInfoWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return CostmapFilterInfoArrayWrapper;
  }
}
module.exports = CostmapFilterInfoWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "header",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "std_msgs",
        "type": "Header",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "type",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "uint8",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "filter_mask_topic",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "string",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "base",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    },
    {
      "name": "multiplier",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "float32",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "CostmapFilterInfo",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "CostmapFilterInfo"
}
*/
