// rclnodejs/generated/nav2_msgs/nav2_msgs__srv__IsPathValid_Request.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PathWrapper = require('../../generated/nav_msgs/nav_msgs__msg__Path.js');
const IsPathValid_RequestRefStruct = StructType({
  path: PathWrapper.refObjectType,
});
const IsPathValid_RequestRefArray = ArrayType(IsPathValid_RequestRefStruct);
const IsPathValid_RequestRefStructArray = StructType({
  data: IsPathValid_RequestRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class IsPathValid_RequestWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new IsPathValid_RequestRefStruct(other._refObject.toObject());
      this._wrapperFields.path =  new PathWrapper(other._wrapperFields.path);
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new IsPathValid_RequestRefStruct();
    this._wrapperFields.path = new PathWrapper();
  }
  static createFromRefObject(refObject) {
    let self = new IsPathValid_RequestWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new IsPathValid_RequestArrayWrapper;
  }
  static get ArrayType() {
    return IsPathValid_RequestArrayWrapper;
  }
  static get refObjectArrayType() {
    return IsPathValid_RequestRefStructArray
  }
  static get refObjectType() {
    return IsPathValid_RequestRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
    }
    this._wrapperFields.path.freeze(own, checkConsistency);
    this._refObject.path = this._wrapperFields.path.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.path.copyRefObject(refObject.path);
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PathWrapper.freeStruct(refObject.path);
  }
  static destoryRawROS(msg) {
    IsPathValid_RequestWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'srv', interfaceName: 'IsPathValid_Request'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get path() {
    return this._wrapperFields.path;
  }
  set path(value) {
    if (value instanceof PathWrapper) {
      this._wrapperFields.path.copy(value);
    } else {
      this._wrapperFields.path.copy(new PathWrapper(value));
    }
  }
  copyRefObject(refObject) {
    this._refObject = new IsPathValid_RequestRefStruct(refObject.toObject());
    this._wrapperFields.path.copyRefObject(this._refObject.path);
  }
  copy(other) {
    this._refObject = new IsPathValid_RequestRefStruct(other._refObject.toObject());
    this._wrapperFields.path.copy(other._wrapperFields.path);
  }
  static get classType() {
    return IsPathValid_RequestWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"Path","stringUpperBound":null,"isPrimitiveType":false},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"IsPathValid_Request","stringUpperBound":null,"isPrimitiveType":false},"msgName":"IsPathValid_Request"};
  }
  hasMember(name) {
    let memberNames = ["path"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class IsPathValid_RequestArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof IsPathValid_RequestWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new IsPathValid_RequestWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IsPathValid_RequestArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to IsPathValid_RequestArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new IsPathValid_RequestRefArray(size);
    this._refObject = new IsPathValid_RequestRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new IsPathValid_RequestWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof IsPathValid_RequestArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "IsPathValid_RequestArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      IsPathValid_RequestWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return IsPathValid_RequestWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return IsPathValid_RequestArrayWrapper;
  }
}
module.exports = IsPathValid_RequestWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_msgs",
        "type": "Path",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "IsPathValid_Request",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "IsPathValid_Request"
}
*/
