// rclnodejs/generated/nav2_msgs/nav2_msgs__action__SmoothPath_Result.js
// This file is automatically generated by Intel rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
const ref = require('@rclnodejs/ref-napi');
const StructType = require('@rclnodejs/ref-struct-di')(ref);
const ArrayType = require('@rclnodejs/ref-array-di')(ref);
const primitiveTypes = require('../../rosidl_gen/primitive_types.js');
const deallocator = require('../../rosidl_gen/deallocator.js');
const translator = require('../../rosidl_gen/message_translator.js');
const PathWrapper = require('../../generated/nav_msgs/nav_msgs__msg__Path.js');
const DurationWrapper = require('../../generated/builtin_interfaces/builtin_interfaces__msg__Duration.js');
const SmoothPath_ResultRefStruct = StructType({
  path: PathWrapper.refObjectType,
  smoothing_duration: DurationWrapper.refObjectType,
  was_completed: primitiveTypes.bool,
});
const SmoothPath_ResultRefArray = ArrayType(SmoothPath_ResultRefStruct);
const SmoothPath_ResultRefStructArray = StructType({
  data: SmoothPath_ResultRefArray,
  size: ref.types.size_t,
  capacity: ref.types.size_t
});
// Define the wrapper class.
class SmoothPath_ResultWrapper {
  constructor(other) {
    this._wrapperFields = {};
    if (typeof other === 'object' && other._refObject) {
      this._refObject = new SmoothPath_ResultRefStruct(other._refObject.toObject());
      this._wrapperFields.path =  new PathWrapper(other._wrapperFields.path);
      this._wrapperFields.smoothing_duration =  new DurationWrapper(other._wrapperFields.smoothing_duration);
      this._was_completedIntialized = true;
    } else if (typeof other !== 'undefined') {
      this._initMembers();
      translator.constructFromPlanObject(this, other);
    } else {
      this._initMembers();
    }
    this.freeze();
  }
  _initMembers() {
    this._refObject = new SmoothPath_ResultRefStruct();
    this._wrapperFields.path = new PathWrapper();
    this._wrapperFields.smoothing_duration = new DurationWrapper();
    this._was_completedIntialized = false;
  }
  static createFromRefObject(refObject) {
    let self = new SmoothPath_ResultWrapper();
    self.copyRefObject(refObject);
    return self;
  }
  static createArray() {
    return new SmoothPath_ResultArrayWrapper;
  }
  static get ArrayType() {
    return SmoothPath_ResultArrayWrapper;
  }
  static get refObjectArrayType() {
    return SmoothPath_ResultRefStructArray
  }
  static get refObjectType() {
    return SmoothPath_ResultRefStruct;
  }
  toRawROS() {
    this.freeze(true);
    return this._refObject.ref();
  }
  freeze(own = false, checkConsistency = false) {
    if (checkConsistency) {
      if (!this._was_completedIntialized) {
        throw new TypeError('Invalid argument: was_completed in SmoothPath_Result');
      }
    }
    this._wrapperFields.path.freeze(own, checkConsistency);
    this._refObject.path = this._wrapperFields.path.refObject;
    this._wrapperFields.smoothing_duration.freeze(own, checkConsistency);
    this._refObject.smoothing_duration = this._wrapperFields.smoothing_duration.refObject;
  }
  serialize() {
    this.freeze(false, true);
    return this._refObject.ref();
  }
  deserialize(refObject) {
    this._wrapperFields.path.copyRefObject(refObject.path);
    this._wrapperFields.smoothing_duration.copyRefObject(refObject.smoothing_duration);
    this._was_completedIntialized = true;
    this._refObject.was_completed = refObject.was_completed;
  }
  toPlainObject(enableTypedArray) {
    return translator.toPlainObject(this, enableTypedArray);
  }
  static freeStruct(refObject) {
    PathWrapper.freeStruct(refObject.path);
    DurationWrapper.freeStruct(refObject.smoothing_duration);
  }
  static destoryRawROS(msg) {
    SmoothPath_ResultWrapper.freeStruct(msg.refObject);
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'SmoothPath_Result'};
  }
  static isPrimitive() {
    return false;
  }
  static get isROSArray() {
    return false;
  }
  get refObject() {
    return this._refObject;
  }
  get path() {
    return this._wrapperFields.path;
  }
  set path(value) {
    if (value instanceof PathWrapper) {
      this._wrapperFields.path.copy(value);
    } else {
      this._wrapperFields.path.copy(new PathWrapper(value));
    }
  }
  get smoothing_duration() {
    return this._wrapperFields.smoothing_duration;
  }
  set smoothing_duration(value) {
    if (value instanceof DurationWrapper) {
      this._wrapperFields.smoothing_duration.copy(value);
    } else {
      this._wrapperFields.smoothing_duration.copy(new DurationWrapper(value));
    }
  }
  get was_completed() {
    if (!this._was_completedIntialized) {
      return undefined;
    }
    return this._refObject.was_completed;
  }
  set was_completed(value) {
    this._was_completedIntialized = true;
    this._refObject.was_completed = value;
  }
  copyRefObject(refObject) {
    this._refObject = new SmoothPath_ResultRefStruct(refObject.toObject());
    this._wrapperFields.path.copyRefObject(this._refObject.path);
    this._wrapperFields.smoothing_duration.copyRefObject(this._refObject.smoothing_duration);
    this._was_completedIntialized = true;
  }
  copy(other) {
    this._refObject = new SmoothPath_ResultRefStruct(other._refObject.toObject());
    this._wrapperFields.path.copy(other._wrapperFields.path);
    this._wrapperFields.smoothing_duration.copy(other._wrapperFields.smoothing_duration);
    this._was_completedIntialized = true;
  }
  static get classType() {
    return SmoothPath_ResultWrapper;
  }
  static get ROSMessageDef() {
    return {"constants":[],"fields":[{"name":"path","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"nav_msgs","type":"Path","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"smoothing_duration","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":"builtin_interfaces","type":"Duration","stringUpperBound":null,"isPrimitiveType":false},"default_value":null},{"name":"was_completed","type":{"isArray":false,"arraySize":null,"isUpperBound":false,"isDynamicArray":false,"isFixedSizeArray":false,"pkgName":null,"type":"bool","stringUpperBound":null,"isPrimitiveType":true},"default_value":null}],"baseType":{"pkgName":"nav2_msgs","type":"SmoothPath_Result","stringUpperBound":null,"isPrimitiveType":false},"msgName":"SmoothPath_Result"};
  }
  hasMember(name) {
    let memberNames = ["path","smoothing_duration","was_completed"];
    return memberNames.indexOf(name) !== -1;
  }
}
// Define the wrapper of array class.
class SmoothPath_ResultArrayWrapper {
  constructor(size = 0) {
    this._resize(size);
  }
  toRawROS() {
    return this._refObject.ref();
  }
  fill(values) {
    const length = values.length;
    this._resize(length);
    values.forEach((value, index) => {
      if (value instanceof SmoothPath_ResultWrapper) {
        this._wrappers[index].copy(value);
      } else {
        this._wrappers[index] = new SmoothPath_ResultWrapper(value);
      }
    });
  }
  // Put all data currently stored in `this._wrappers` into `this._refObject`
  freeze(own) {
    this._wrappers.forEach((wrapper, index) => {
      wrapper.freeze(own);
      this._refArray[index] = wrapper.refObject;
    });
    this._refObject.size = this._wrappers.length;
    this._refObject.capacity = this._wrappers.length;
    if (this._refObject.capacity === 0) {
      this._refObject.data = null
    } else {
    this._refObject.data = this._refArray.buffer;
    }
  }
  get refObject() {
    return this._refObject;
  }
  get data() {
    return this._wrappers;
  }
  get size() {
    return this._wrappers.length;
  }
  set size(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SmoothPath_ResultArrayWrapper.size setter');
      return;
    }
    return this._resize(value);
  }
  get capacity() {
    return this._wrappers.length;
  }
  set capacity(value) {
    if (typeof value != 'number') {
      throw new TypeError('Invalid argument: should provide a number to SmoothPath_ResultArrayWrapper.capacity setter');
    }
    return this._resize(value);
  }
  get refObject() {
    return this._refObject;
  }
  _resize(size) {
    if (size < 0) {
      throw new RangeError('Invalid argument: should provide a positive number');
      return;
    }
    this._refArray = new SmoothPath_ResultRefArray(size);
    this._refObject = new SmoothPath_ResultRefStructArray();
    this._refObject.size = size;
    this._refObject.capacity = size;
    this._wrappers = new Array();
    for (let i = 0; i < size; i++) {
      this._wrappers.push(new SmoothPath_ResultWrapper());
    }
  }
  // Copy all data from `this._refObject` into `this._wrappers`
  copyRefObject(refObject) {
    this._refObject = refObject;
    let refObjectArray = this._refObject.data;
    refObjectArray.length = this._refObject.size;
    this._resize(this._refObject.size);
    for (let index = 0; index < this._refObject.size; index++) {
      this._wrappers[index].copyRefObject(refObjectArray[index]);
    }
  }
  copy(other) {
    if (! (other instanceof SmoothPath_ResultArrayWrapper)) {
      throw new TypeError('Invalid argument: should provide "SmoothPath_ResultArrayWrapper".');
    }
    this._resize(other.size);
    // Array deep copy
    other._wrappers.forEach((wrapper, index) => {
      this._wrappers[index].copy(wrapper);
    });
  }
  static freeArray(refObject) {
    let refObjectArray = refObject.data;
    refObjectArray.length = refObject.size;
    for (let index = 0; index < refObject.size; index++) {
      SmoothPath_ResultWrapper.freeStruct(refObjectArray[index]);
    }
  }
  static get elementType() {
    return SmoothPath_ResultWrapper;
  }
  static get isROSArray() {
    return true;
  }
  static get useTypedArray() {
    return false;
  }
  get classType() {
    return SmoothPath_ResultArrayWrapper;
  }
}
module.exports = SmoothPath_ResultWrapper;
/*
 * The following is the original spec object coming from parser:
{
  "constants": [],
  "fields": [
    {
      "name": "path",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "nav_msgs",
        "type": "Path",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "smoothing_duration",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": "builtin_interfaces",
        "type": "Duration",
        "stringUpperBound": null,
        "isPrimitiveType": false
      },
      "default_value": null
    },
    {
      "name": "was_completed",
      "type": {
        "isArray": false,
        "arraySize": null,
        "isUpperBound": false,
        "isDynamicArray": false,
        "isFixedSizeArray": false,
        "pkgName": null,
        "type": "bool",
        "stringUpperBound": null,
        "isPrimitiveType": true
      },
      "default_value": null
    }
  ],
  "baseType": {
    "pkgName": "nav2_msgs",
    "type": "SmoothPath_Result",
    "stringUpperBound": null,
    "isPrimitiveType": false
  },
  "msgName": "SmoothPath_Result"
}
*/
