// rclnodejs/generated/nav2_msgs/nav2_msgs__action__DriveOnHeading.js
// This file is automatically generated by rclnodejs
//
// *** DO NOT EDIT directly
//
'use strict';
class nav2_msgs__action__DriveOnHeading {
  static get Goal() {
    return require('./nav2_msgs__action__DriveOnHeading_Goal.js');
  }
  static get Result() {
    return require('./nav2_msgs__action__DriveOnHeading_Result.js');
  }
  static get Feedback() {
    return require('./nav2_msgs__action__DriveOnHeading_Feedback.js');
  }
  static get impl() {
    return {
      get SendGoalService() {
        return require('./nav2_msgs__action__DriveOnHeading_SendGoal.js');
      },
      get GetResultService() {
        return require('./nav2_msgs__action__DriveOnHeading_GetResult.js');
      },
      get FeedbackMessage() {
        return require('./nav2_msgs__action__DriveOnHeading_FeedbackMessage.js');
      },
      get CancelGoal() {
        return require('../action_msgs/action_msgs__srv__CancelGoal.js');
      },
      get GoalStatusArray() {
        return require('../action_msgs/action_msgs__msg__GoalStatusArray.js');
      }
    };
  }
  static type() {
    return {pkgName: 'nav2_msgs', subFolder: 'action', interfaceName: 'DriveOnHeading'};
  }
}
module.exports = nav2_msgs__action__DriveOnHeading;
